lib/client/conntest/ssh.go

```diff
@@ -49,30 +59,239 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
-	id := uuid.NewString()
-	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(id, map[string]string{},
+	if req.ResourceKind != types.KindNode {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindNode, req.ResourceKind)
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(connectionDiagnosticID, map[string]string{},
 		types.ConnectionDiagnosticSpecV1{
-			Message: types.DiagnosticMessageWaiting,
+			Message: types.DiagnosticMessageFailed,
 		})
+	if err != nil {
+		return nil, trace.Wrap(err, "new conn diag")
+	}
+
+	if err := s.clt.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err, "create conn diag")
+	}
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return nil, trace.Wrap(err, "generate rsa key")
+	}
+
+	currentUser, err := s.clt.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err, "current user")
+	}
+
+	certs, err := s.clt.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: connectionDiagnosticID,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.Cert = certs.SSH
+	key.TLSCert = certs.TLS
+
+	certAuths, err := s.clt.GetCertAuthorities(ctx, types.HostCA, false /* loadKeys */)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	if err := s.cd.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+	hostkeyCallback, err := hostkeyCallbackFromCAs(certAuths)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+
+	keyAuthMethod, err := key.AsAuthMethod()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName, err := s.clt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientConfTLS, err := key.TeleportClientTLSConfig(nil, []string{clusterName.GetClusterName()})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	host := req.ResourceName
+	tlsRoutingEnabled := true
+
+	webProxyAddr, sshProxyAddr, err := s.proxyAddrs()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.KeyIndex = client.KeyIndex{
+		Username:    req.SSHPrincipal,
+		ProxyHost:   webProxyAddr,
+		ClusterName: clusterName.GetClusterName(),
+	}
+
+	processStdout := &bytes.Buffer{}
+
+	clientConf := client.MakeDefaultConfig()
+	clientConf.AuthMethods = []ssh.AuthMethod{keyAuthMethod}
+	clientConf.Host = host
+	clientConf.HostKeyCallback = hostkeyCallback
+	clientConf.HostLogin = req.SSHPrincipal
+	clientConf.SkipLocalAuth = true
+	clientConf.SSHProxyAddr = sshProxyAddr // TODO(marco): remove the next line?
+	clientConf.SSHProxyAddr = webProxyAddr
```

Comment:
This should be set to web proxy address if TLS routing is enabled and to SSH proxy address otherwise.